programs/mango-v4/src/instructions/perp_liq_base_or_positive_pnl.rs

```diff
@@ -230,163 +238,294 @@ pub(crate) fn liquidation_action(
-    let mut current_expected_health =
-        liqee_liq_end_health + current_expected_perp_health - initial_weighted_perp_health;
+    let mut pnl_transfer = I80F48::ZERO;
+    let mut current_uhupnl = perp_info.unweighted_health_unsettled_pnl(liq_end_type);
+    let mut current_health = liqee_liq_end_health;
+    let mut current_settle_token = settle_token_balance.spot_and_perp;
 
+    // Helper function to reduce base position in exchange for effective settle token balance changes
     let mut reduce_base = |step: &str,
-                           health_amount: I80F48,
-                           health_per_lot: I80F48,
-                           current_unweighted_perp_health: &mut I80F48| {
-        // How much are we willing to increase the unweighted perp health?
-        let health_limit = health_amount
-            .min(-current_expected_health)
-            .max(I80F48::ZERO);
+                           uhupnl_limit: I80F48,
+                           // How much effective settle token will be gained when taking future
+                           // upnl settlement into account.
+                           expected_settle_token_per_lot: I80F48,
+                           // Effective settle token gained directly from the operation (before settlement)
+                           actual_settle_token_per_lot: I80F48,
+                           uhupnl_per_lot: I80F48,
+                           current_uhupnl: &mut I80F48,
+                           current_settle_token: &mut I80F48,
+                           current_health: &mut I80F48| {
+        let max_settle_token_for_health = spot_amount_given_for_health_zero(
+            *current_health,
+            *current_settle_token,
+            settle_token_info.asset_weighted_price(liq_end_type),
+            settle_token_info.liab_weighted_price(liq_end_type),
+        )
+        .unwrap();
+
+        let max_settle_token = max_settle_token_for_health.min(uhupnl_limit);
+
         // How many lots to transfer?
-        let base_lots = (health_limit / health_per_lot)
+        let base_lots = (max_settle_token / expected_settle_token_per_lot)
             .ceil() // overshoot to aim for init_health >= 0
             .to_num::<i64>()
             .min(liqee_base_lots.abs() - base_reduction)
             .min(max_base_transfer.abs() - base_reduction)
             .max(0);
-        let unweighted_change = I80F48::from(base_lots) * unweighted_health_per_lot;
-        let current_unweighted = *current_unweighted_perp_health;
-        let new_unweighted_perp = current_unweighted + unweighted_change;
-        let new_expected_perp = expected_perp_health(new_unweighted_perp);
-        let new_expected_health =
-            current_expected_health + (new_expected_perp - current_expected_perp_health);
+
+        let expected_settle_token_gain = expected_settle_token_per_lot * I80F48::from(base_lots);
```

Comment:
are the expected_* variables just for logging?

Reply:
Yes. It used to be different. I'm not sure it's worth keeping them for logging purposes. What do you think?

Reply:
honestly, I need to look at some real world liq logs to see how the logging in this ix is, so lets revisit once we have some real world examples

Reply:
ok